### PR TITLE
Support AdditionalProperties

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/autorest.csharp/issues/3454

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
 
-using System;
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Runtime.CompilerServices;
 
 namespace AutoRest.CSharp.Common.Input
 {
-    internal record InputModelType(string Name, string? Namespace, string? Accessibility, string? Deprecated, string? Description, InputModelTypeUsage Usage, IReadOnlyList<InputModelProperty> Properties, InputModelType? BaseModel, IReadOnlyList<InputModelType> DerivedModels, string? DiscriminatorValue, string? DiscriminatorPropertyName, bool IsNullable)
+    internal record InputModelType(string Name, string? Namespace, string? Accessibility, string? Deprecated, string? Description, InputModelTypeUsage Usage, IReadOnlyList<InputModelProperty> Properties, InputModelType? BaseModel, IReadOnlyList<InputModelType> DerivedModels, string? DiscriminatorValue, string? DiscriminatorPropertyName, InputDictionaryType? InheritedDictionaryType, bool IsNullable)
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:10]: this change is taken from [Alexander's PR](https://github.com/Azure/autorest.csharp/pull/3726).

In most cases, it should work perfectly fine that we could get `InheritedDictionaryType` from the `BaseModel` property, especially in the scenario of using typespec, because in typespec, a model could only have one base model.
This is a backward compat thing because in swagger, a model could have multiple base models, and this property is not null when any of thme is a dictionary

Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:10]: We could consider change this to a calculated property, for instance, for the case we have now, it could simply be:
```csharp
internal record InputModelType(/*parameters unchanged*/)
{
public InputDictionaryType InheritDictionaryType => BaseModel as InputDictionaryType;
}
```
and update the implementation here when we support the full conversion from swagger to input types.

But as for now, to reduce the amount of changes I will just leave it here, and may go over this again when the feature is done.

Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:10]: And we cannot just make the InputDictionaryType the base model - because BaseModel only accepts `InputModelType`.

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts [hunk 2]
```diff
@@ -347,12 +341,12 @@ export function getInputType(
 
     function getInputModelType(m: Model): InputType {
         /* Array and Map Type. */
-        if (!isNeverType(m)) {
```
Comment (ArcturusZhang) [inline src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:350]: we only get into this function when `m.kind === "Model"`, and the `isNeverType` is checking `m.kind === "intrinsic"` which will never be true, therefore here we do not need to check this.

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts [hunk 3]
```diff
@@ -558,14 +554,15 @@ export function getInputType(
...
+                Description: getDoc(program, m) ?? "",
                 IsNullable: false,
                 DiscriminatorPropertyName: discriminator?.propertyName,
                 DiscriminatorValue: getDiscriminatorValue(m, baseModel),
                 Usage: Usage.None,
+                InheritedDictionaryType: inheritedDictionaryType, // InheritedDictionaryType represent the type of additional properties property
                 BaseModel: baseModel, // BaseModel should be the last but one assigned to model
                 Properties: properties // Properties should be the last assigned to model
-            } as InputModelType;
+            };
```
Comment (ArcturusZhang) [inline src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:565]: `as` kind of by-passes the static type validation.
For instance, here the variable `model` has a type of `InputModelType`, which requires a string for `Description`. But here we are creating an object, with `Description` of type `string | undefined`.
Without the `as`, this is a compilation error, if we have the `as` statement, the error is actually suppressed.
We should not do that, therefore I removed it, and assign empty string if description is empty.

File: test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/Models/ChatCompletionFunctions.cs [hunk 4]
```diff
@@ -24,7 +23,7 @@ public partial class ChatCompletionFunctions
         /// `{\"type\": \"object\", \"properties\": {}}`.
         /// </param>
         /// <exception cref="ArgumentNullException"> <paramref name="name"/> or <paramref name="parameters"/> is null. </exception>
-        public ChatCompletionFunctions(string name, IDictionary<string, BinaryData> parameters)
+        public ChatCompletionFunctions(string name, ChatCompletionFunctionParameters parameters)
```
Comment (ArcturusZhang) [inline test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/Models/ChatCompletionFunctions.cs:26]: this changes because in its [tsp definition](https://github.com/Azure/autorest.csharp/blob/5ed56117b265c61877a8c2b8b9892266071f1f0c/test/UnbrandedProjects/Platform-OpenAI-TypeSpec/completions/models.tsp#L204) we have:
```
model ChatCompletionFunctionParameters is Record<unknown>;
```
when we do not have additional properties, we parse this as a dictionary, but when we support additional properties, this should be a model with additional properties property, therefore we have this change.

Comment (ArcturusZhang) [reply test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/Models/ChatCompletionFunctions.cs:26]: and if they only want a dictionary in their tsp, they could 
1. just use plain `Record<unknown>` on the property
2. use alias like this:
```
alias ChatCompletionFunctionParameters = Record<unknown>;
```

File: src/TypeSpec.Extension/Emitter.Csharp/test/Unit/modelType.test.ts [hunk 5]
```diff
@@ -76,7 +76,6 @@ op test(@body input: Pet): Pet;
                         Kind: "String",
                         IsNullable: false
                     },
-                    IsNullable: false,
```
Comment (ArcturusZhang) [inline src/TypeSpec.Extension/Emitter.Csharp/test/Unit/modelType.test.ts:79]: this is removed because we never have a `IsNullable` property in the type `InputModelProperty`.

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts [hunk 6]
```diff
@@ -667,40 +666,65 @@ export function getInputType(
...
                 `No specified type for discriminator property '${model.DiscriminatorPropertyName}'. Assume it is a string.`
             );
-            const discriminatorProperty = {
+            const discriminatorProperty: InputModelProperty = {
                 Name: model.DiscriminatorPropertyName,
                 SerializedName: model.DiscriminatorPropertyName,
                 Description: "Discriminator",
                 IsRequired: true,
                 IsReadOnly: false,
-                IsNullable: false,
```
Comment (ArcturusZhang) [inline src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:676]: this is removed because we never have a `IsNullable` property in `InputModelProperty` type.

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts [hunk 7]
```diff
@@ -667,40 +666,65 @@ export function getInputType(
...
-            return undefined;
+    function getInputModelBaseType(
+        m: Model
+    ): [InputModelType | undefined, InputDictionaryType | undefined] {
+        const baseModel = m.baseModel;
+        const sourceModel = m.sourceModel;
+
+        // we cannot have both `extends` and `is`, therefore only one of baseModel and sourceModel can be defined
+        if (
+            sourceModel !== undefined &&
```
Comment (chunyu3) [inline src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:695]:
```suggestion
            !sourceModel &&
```

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts [hunk 8]
```diff
@@ -667,40 +666,65 @@ export function getInputType(
...
+                    sourceModel.indexer.key,
+                    sourceModel.indexer.value
+                )
+            ];
         }
 
-        // TypeSpec "primitive" types can't be base types for models
-        if (program.checker.isStdType(m)) {
-            return undefined;
+        if (baseModel !== undefined) {
```
Comment (chunyu3) [inline src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:707]:
```suggestion
        if (!baseModel) {
```

**Review by chunyu3 — APPROVED** [review]
Approved with some change request.

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts (hunk 7 above)
Comment (ArcturusZhang) [reply src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:695]: well when I wrote this, I kind of did not figure out how to simplify this expression and ended up writing this
I will update

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts [hunk 9]
```diff
@@ -667,40 +666,62 @@ export function getInputType(
...
-        if (!m) {
-            return undefined;
-        }
 
-        // Arrays and dictionaries can't be a base type
-        if (m.indexer) {
-            return undefined;
+    function getInputModelBaseType(
+        m: Model
+    ): [InputModelType | undefined, InputDictionaryType | undefined] {
```
Comment (pshao25) [inline src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:689]: Just curious, could array be a base type?

Comment (lirenhe) [inline src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:689]: Can we return the union of  'InputModelType |  InputDictionaryType | undefined' for this method?

File: src/TypeSpec.Extension/Emitter.Csharp/src/type/inputType.ts [hunk 10]
```diff
@@ -62,7 +63,7 @@ export interface InputEnumType extends InputType {
     Namespace?: string;
     Accessibility?: string;
     Deprecated?: string;
-    Description: string;
+    Description?: string;
```
Comment (lirenhe) [inline src/TypeSpec.Extension/Emitter.Csharp/src/type/inputType.ts:66]: Is this change related to the support of addtionalProperties?

File: src/TypeSpec.Extension/Emitter.Csharp/test/Unit/modelType.test.ts [hunk 11]
```diff
@@ -76,7 +76,6 @@ op test(@body input: Pet): Pet;
                         Kind: "String",
                         IsNullable: false
```
Comment (lirenhe) [inline src/TypeSpec.Extension/Emitter.Csharp/test/Unit/modelType.test.ts:77]: Can we have some tests in Emitter to cover  additionalProperties?

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts (hunk 9 above)
Comment (ArcturusZhang) [reply src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:689]: No. We tried in playground, it is OK to write this:
```
model Foo extends Array<string> {}
```
but Timothe confirms this is a bug.

File: test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/Models/ChatCompletionFunctions.Serialization.cs [hunk 12]
```diff
@@ -21,25 +21,7 @@ void IUtf8JsonWriteable.Write(Utf8JsonWriter writer)
                 writer.WriteStringValue(Description);
             }
             writer.WritePropertyName("parameters"u8);
-            writer.WriteStartObject();
-            foreach (var item in Parameters)
-            {
-                writer.WritePropertyName(item.Key);
-                if (item.Value == null)
```
Comment (lirenhe) [inline test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/Models/ChatCompletionFunctions.Serialization.cs:28]: Seems the null value handling was removed in the serialization of additionalProperties, is this expected?

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts (hunk 9 above)
Comment (ArcturusZhang) [reply src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts:689]: Then we will have to write this in the place where we call this:
```
if (baseModel === undefined)
{
// undefined case
}
else if (isModelType(baseModel))
{
// do base model
}
else if (isInputDictionaryType(baseModel))
{
// do inherited dictionary type
}
```
I believe this looks way more ugly.

File: src/TypeSpec.Extension/Emitter.Csharp/src/type/inputType.ts (hunk 10 above)
Comment (ArcturusZhang) [reply src/TypeSpec.Extension/Emitter.Csharp/src/type/inputType.ts:66]: No this is a clean up to align with the definitions.

File: src/TypeSpec.Extension/Emitter.Csharp/test/Unit/modelType.test.ts (hunk 11 above)
Comment (ArcturusZhang) [reply src/TypeSpec.Extension/Emitter.Csharp/test/Unit/modelType.test.ts:77]: Sure, I will add some

File: test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/Models/ChatCompletionFunctions.Serialization.cs (hunk 12 above)
Comment (ArcturusZhang) [reply test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/Models/ChatCompletionFunctions.Serialization.cs:28]: This change is expected. But `the null value handling was removed in the serialization of additionalProperties` is not the correct way to understand this line of change.
This model has a property `Parameters`. Previously, we have:
```csharp
class ChatCompletionFunctions
{
public IDictionary<string, BinaryData> Parameters {get;}
}
```
and now we have:
```csharp
class ChatCompletionFunctions
{
public ChatCompletionFunctionParameters Parameters {get;}
}
```
and `ChatCompletionFunctionParameters` is a new model, therefore this bunch of code changes to just `WriteObjectValue` which calls the serialization of the model.

Comment (ArcturusZhang) [reply test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/Models/ChatCompletionFunctions.Serialization.cs:28]: I see. The serialization of `AdditionalProperties` of type `IDictionary<string, BinaryData>` whihc is the same as we changed before, but their serialization are different.
The serialization for additional properties is missing a `item == null` check.
I will not change it on this PR because other libraries have a lot of use cases of additional properties property, changing it causes a lot change to this PR.
I will follow up on this issue after this merges.

File: src/TypeSpec.Extension/Emitter.Csharp/test/Unit/modelType.test.ts (hunk 11 above)
Comment (ArcturusZhang) [reply src/TypeSpec.Extension/Emitter.Csharp/test/Unit/modelType.test.ts:77]: tests added

**Review by lirenhe — APPROVED** [review]
